languageServer/src/TextDocumentWrapper.ts

```diff
@@ -0,0 +1,294 @@
+
+    const completions: CompletionItem[] = []
+
+    for (const { label, body } of this.getCompletionTemplates()) {
+      const completionDocument: Document = new Document(body)
+      const newSource = completionDocument.toString()
+      const item = CompletionItem.create(label)
+      item.insertTextFormat = InsertTextFormat.Snippet
+      item.insertText = newSource
+      completions.push(item)
+    }
+
+    return completions
+  }
+
+  public getCodeActions(params: CodeActionParams): CodeAction[] | null {
+    if (!this.uri.endsWith('dvc.yaml')) {
+      return null
+    }
+
+    const codeActions: CodeAction[] = []
+
+    const model = new DvcYaml(this)
+    const stage = model.getStageAt(params.range.start)
+    if (stage?.cmd) {
+      const filesUsed = stage.cmd.getReferencedFiles()
+      const missingDeps = filesUsed.filter(path => !stage.deps?.has(path))
+
+      if (missingDeps.length > 0) {
+        stage.addDependencies(missingDeps)
+        const yamlSource = stage.parentDvcYaml.toString()
+        const textEdits = this.calculateTextEdits(yamlSource)
+
+        const workspaceEdit: WorkspaceEdit = {
+          changes: {
+            [this.uri]: textEdits
+          }
+        }
+
+        const codeAction = CodeAction.create(
+          'Add cmd files as dependencies',
+          workspaceEdit,
+          CodeActionKind.RefactorRewrite
+        )
+
+        codeActions.push(codeAction)
+      }
+    }
+
+    return codeActions
+  }
+
+  public getDefinitions(position: Position): Location[] {
+    const theSymbol = this.symbolAt(position)
+    if (theSymbol) {
+      const allDocs = this.documents.all()
+
+      const locationsAccumulator = []
+
+      for (const txtDoc of allDocs) {
+        const finder = this.createFinder(txtDoc)
+        const locations = finder.findLocationsFor(theSymbol)
+        locationsAccumulator.push(...locations)
+      }
+
+      return locationsAccumulator ?? []
+    }
+    return []
+  }
+
+  private getCompletionTemplates() {
+    const completionTemplates: CompletionTemplate[] = [
+      {
+        body: {
+          stages: { $1: { cmd: '$2' } }
+        },
+        label: 'stages'
+      },
+      {
+        body: {
+          $1: { cmd: '$2' }
+        },
+        label: 'Add stage'
+      }
+    ]
+    for (const path of this.pythonFilePaths) {
+      completionTemplates.push({
+        body: {
+          cmd: `python ${path} $1`
+        },
+        label: `cmd: ${path}`
+      })
+    }
+
+    return completionTemplates
+  }
+
+  private getTemplateExpressionSymbolsInsideScalar(
+    scalarValue: string,
+    nodeOffset: number
+  ) {
+    const templateSymbols: DocumentSymbol[] = []
+
+    const templates = scalarValue.matchAll(variableTemplates)
+    for (const template of templates) {
+      const expression = template[1]
+      const expressionOffset: number = nodeOffset + (template.index ?? 0) + 2 // To account for the '${'
+      const symbols = expression.matchAll(/[\dA-Za-z]+/g)
```

Comment:
Looks a lot like the regex [here](https://github.com/iterative/vscode-dvc/pull/2359/files#diff-f6824a6c404b5ff658c1c62a45c9ea9cb5a571cab1951e3adfa52d64174144ebR54). Same comment about trying to explain what you're matching and why